File: Algorithm/C/Sort/ShellSort.c
/*
  希尔排序（数据结构与算法分析P168）;
*/
void ShellSort(int A[], int N)
{
  int Increment,i,j;
  int Tmp;
  for(Increment = N/2; Increment > 0; Increment/=2)
    {
      //相当于一个间隔为Increment的插入排序
      for(i = Increment; i < N; i++)
	{
	  Tmp = A[i];
	  for(j = i;j >= Increment && Tmp < A[j-Increment];j-=Increment)
	    A[j] = A[j-Increment];
	  A[j] = Tmp;
	}
    }
  printf("\n");
}

File: Algorithm/python/sort/shell_sort.py
from unittest import TestCase
from random import sample


class TestSort(TestCase):
    def setUp(self):
        self.list = sample(range(0, 1000), 20)

    def test_bubble_sort(self):
        sorted_list = shell_sort(self.list)
        for i in range(0, len(sorted_list) - 1):
            self.assertTrue(sorted_list[i] < sorted_list[i + 1])


def shell_sort(numbers):
    increment = len(numbers) // 2
    while increment > 0:
        for i in range(0, increment):
            sub_list = list(range(i, len(numbers), increment))
            for j in range(1, len(sub_list)):
                insert_number = numbers[sub_list[j]]
                insert_index = j
                while insert_number < numbers[sub_list[insert_index - 1]] and insert_index > 0:
                    numbers[sub_list[insert_index]] = numbers[sub_list[insert_index - 1]]
                    insert_index -= 1
                numbers[sub_list[insert_index]] = insert_number
        increment //= 2
    return numbers
